octave/fsk_cml_sam.m

```diff
@@ -0,0 +1,275 @@
+
+rx_sig_pow = rx_sig_pow/Nsymbols;
+rx_nse_pow = rx_nse_pow/Nsymbols;
+v_est = sqrt(rx_sig_pow-rx_nse_pow); 
+SNRest = rx_sig_pow/rx_nse_pow;
+sig_est = sqrt(rx_nse_pow/2);    % for Rayleigh: 2nd raw moment = 2 .sigma^2
+Kest = rx_sig_pow/(2.0*sig_est^2) -1.0;
+
+Nerrors = sum(xor(tx_bits, rx_bits));
+raw_ber = Nerrors/Nbits;
+printf('EcNodB: %4.1f  M: %2d Uncoded Nbits: %5d Nerrors: %4d (Raw) BER: %1.3f\n', ...
+    EcNodB, M, Nbits, Nerrors, raw_ber);
+if plt==1, plot_hist(rx_filt,tx_symbols, M); end
+
+end
+
+%%%%%%%%%%%%%%%%%%%%%%%%%
+function [Nerrors raw_ber EcNodB] = run_single_ldpc(M, Ltype, Nbits,EbNodB, plt, HRA)
+
+disp([num2str(M) 'FSK coded test ...  '])
+if M==2
+    bps = 1; modulation = 'FSK'; mod_order=2; mapping = 'gray';
+elseif M==4
+    bps = 2; modulation = 'FSK'; mod_order=4; mapping = 'gray';
+else
+    error('sorry - bad value of M!');
+end
+decoder_type = 0; max_iterations = 100;
+
+%load HRAa_1536_512.mat
+%[H_rows, H_cols] = Mat2Hrows(HRA);
+%code_param.H_rows = H_rows;
+%code_param.H_cols = H_cols;
+%code_param.P_matrix = [];
+
+Hsize=size(HRA); 
+Krate = (Hsize(2)-Hsize(1))/Hsize(2) % Krate = 3/4;
+EcNodB = EbNodB + 10*log10(Krate);
+code_param = ldpc_init_user(HRA, modulation, mod_order, mapping);
+Nframes = floor(Nbits/code_param.data_bits_per_frame)
+Nbits = Nframes*code_param.data_bits_per_frame
+
+% Encoder
+data_bits = round(rand(1,code_param.data_bits_per_frame));
+tx_bits = [];
+for f=1:Nframes;
+    codeword_bits = LdpcEncode(data_bits, code_param.H_rows, code_param.P_matrix);
+    tx_bits = [tx_bits codeword_bits];
+end
+%tx_bits = zeros(1,length(tx_bits));
+
+% modem/channel simulation
+[raw_ber rx_filt rx_bits tx_symbols demapper sig_est SNRlin v_est] = run_single(tx_bits,M,EcNodB, plt ); 
+
+% Decoder
+Nerrors = 0;
+for f=1:Nframes
+    st = (f-1)*code_param.coded_bits_per_frame/bps + 1;
+    en = st + code_param.coded_bits_per_frame/bps - 1;
+    
+    if or(Ltype==1, Ltype==3)
+        if bps==1,
+            sd = rx_filt(st:en,1) - rx_filt(st:en,2);
+            %  OR ind = rx_filt(st:en,1) > rx_filt(st:en,2);
+            %     llr = ind'*2 -1;   % this works but use SNR scaling
+            if Ltype==3, HD=1; else, HD = 0;  end
+            llr = sd_to_llr(sd, HD)';
+        end
+        if bps==2,
+            if Ltype==3,
+                llr = mfsk_hd_to_llrs(rx_filt(st:en,:), demapper);
+            else
+                error('Ltype =1 not provided for coded 4FSK');
+            end
+        end
+    end
+    if Ltype==2,     % SDs are converted to LLRs
+        % use the SD amp estimates 
+        if plt==1, plot_pdf(v_est, sig_est);  end
+        llr = mfsk_sd_to_llrs(rx_filt(st:en,:), demapper, v_est, sig_est);
+    end
+    
+    if Ltype==4,
+        % use CML demod:   non-coherent; still seems to need amplitude estimate  
+        symL = DemodFSK(1/v_est*rx_filt(st:en,:)', SNRlin, 1);     
+        llr = -Somap(symL);  % now convert to bit LLRs 
+    end
```

Comment:
OK so these are the CML functions we would need to get going In C.  I think I already have a version of Somap. I have taken a look at the estimators for v_est and SNRLin - so no Genie estimators left - cool :slightly_smiling_face: